File: src/parser.cpp
#include "parser.h"
#include <stdexcept>
#include <iostream>

Parser::Parser(const TokenList& tokenList) : tokens(tokenList), position(0) {}

Token Parser::currentToken() {
    if (position >= tokens.size()) {
        return Token(TokenType::END_OF_FILE, "", 0, 0);
    }
    return tokens[position];
}

Token Parser::peek(int offset) {
    if (position + offset >= tokens.size()) {
        return Token(TokenType::END_OF_FILE, "", 0, 0);
    }
    return tokens[position + offset];
}

void Parser::advance() {
    position++;
}

std::unique_ptr<ASTNode> Parser::parse() {
    return parseProgram();
}

std::unique_ptr<ASTNode> Parser::parseProgram() {
    auto program = std::make_unique<ProgramNode>();
    
    while (currentToken().type != TokenType::END_OF_FILE) {
        try {
            auto statement = parseStatement();
            if (statement) {
                program->statements.push_back(std::move(statement));
            }
        } catch (const std::exception& e) {
            std::cerr << "Parser error at line " << currentToken().line 
                      << ", column " << currentToken().column << ": " << e.what() << std::endl;
            
            // Skip to the next statement to try to recover
            while (currentToken().type != TokenType::EOL && 
                   currentToken().type != TokenType::END_OF_FILE) {
                advance();
            }
            if (currentToken().type == TokenType::EOL) {
                advance();
            }
        }
    }
    
    return program;
}

std::unique_ptr<ASTNode> Parser::parseStatement() {
    // Skip empty lines
    while (currentToken().type == TokenType::EOL) {
        advance();
    }
    
    if (currentToken().type == TokenType::END_OF_FILE) {
        return nullptr;
    }
    
    // Handle different statement types
    switch (currentToken().type) {
        case TokenType::FUNCTION:
            return parseFunction();
        
        case TokenType::VARIABLE:
        case TokenType::CONSTANT:
            return parseVariableDeclaration();
        
        case TokenType::IF:
            return parseIfStatement();
        
        case TokenType::WHILE:
            return parseWhileLoop();
        
        case TokenType::PRINT:
            return parsePrintStatement();
        
        case TokenType::RETURN: {
            advance();
            auto returnValue = parseExpression();
            return std::make_unique<ReturnNode>(std::move(returnValue));
        }
        
        case TokenType::LEFT_BRACE:
            return parseBlock();
        
        default: {
            // Assume it's an expression statement (assignment or function call)
            auto expr = parseExpression();
            return expr;
        }
    }
}

std::unique_ptr<ASTNode> Parser::parseBlock() {
    auto block = std::make_unique<BlockNode>();
    
    // Expect a left brace
    if (currentToken().type != TokenType::LEFT_BRACE) {
        throw std::runtime_error("Expected '{' at the beginning of a block");
    }
    advance();
    
    // Parse statements until closing brace
    while (currentToken().type != TokenType::RIGHT_BRACE && 
           currentToken().type != TokenType::END_OF_FILE) {
        auto statement = parseStatement();
        if (statement) {
            block->statements.push_back(std::move(statement));
        }
    }
    
    // Expect a right brace
    if (currentToken().type != TokenType::RIGHT_BRACE) {
        throw std::runtime_error("Expected '}' at the end of a block");
    }
    advance();
    
    return block;
}

std::unique_ptr<ASTNode> Parser::parseVariableDeclaration() {
    bool isConstant = currentToken().type == TokenType::CONSTANT;
    advance();
    
    if (currentToken().type != TokenType::IDENTIFIER) {
        throw std::runtime_error("Expected identifier after 'var' or 'const'");
    }
    
    std::string varName = currentToken().value;
    advance();
    
    if (currentToken().type != TokenType::ASSIGN) {
        throw std::runtime_error("Expected 'is' after variable name");
    }
    advance();
    
    auto initialValue = parseExpression();
    
    return std::make_unique<VariableDeclarationNode>(varName, isConstant, std::move(initialValue));
}

std::unique_ptr<ASTNode> Parser::parseExpression() {
    auto left = parseTerm();
    
    while (currentToken().type == TokenType::ASSIGN || 
           currentToken().type == TokenType::EQUALS || 
           currentToken().type == TokenType::NOT_EQUALS || 
           currentToken().type == TokenType::GREATER || 
           currentToken().type == TokenType::LESS) {
        
        BinaryOpNode::Op op;
        switch (currentToken().type) {
            case TokenType::ASSIGN:
                op = BinaryOpNode::Op::Assign;
                break;
            case TokenType::EQUALS:
                op = BinaryOpNode::Op::Equals;
                break;
            case TokenType::NOT_EQUALS:
                op = BinaryOpNode::Op::NotEquals;
                break;
            case TokenType::GREATER:
                op = BinaryOpNode::Op::Greater;
                break;
            case TokenType::LESS:
                op = BinaryOpNode::Op::Less;
                break;
            default:
                throw std::runtime_error("Unexpected operator");
        }
        advance();
        
        // For "than" in comparisons like "greater than", "less than"
        if (op == BinaryOpNode::Op::Greater || op == BinaryOpNode::Op::Less) {
            if (currentToken().type == TokenType::IDENTIFIER && currentToken().value == "than") {
                advance();
            }
        }
        
        auto right = parseTerm();
        left = std::make_unique<BinaryOpNode>(op, std::move(left), std::move(right));
    }
    
    return left;
}

std::unique_ptr<ASTNode> Parser::parseTerm() {
    auto left = parseFactor();
    
    while (currentToken().type == TokenType::PLUS || 
           currentToken().type == TokenType::MINUS) {
        
        BinaryOpNode::Op op;
        switch (currentToken().type) {
            case TokenType::PLUS:
                op = BinaryOpNode::Op::Add;
                break;
            case TokenType::MINUS:
                op = BinaryOpNode::Op::Subtract;
                break;
            default:
                throw std::runtime_error("Unexpected operator");
        }
        advance();
        
        auto right = parseFactor();
        left = std::make_unique<BinaryOpNode>(op, std::move(left), std::move(right));
    }
    
    return left;
}

std::unique_ptr<ASTNode> Parser::parseFactor() {
    auto left = parsePrimary();
    
    while (currentToken().type == TokenType::MULTIPLY || 
           currentToken().type == TokenType::DIVIDE) {
        
        BinaryOpNode::Op op;
        switch (currentToken().type) {
            case TokenType::MULTIPLY:
                op = BinaryOpNode::Op::Multiply;
                break;
            case TokenType::DIVIDE:
                op = BinaryOpNode::Op::Divide;
                break;
            default:
                throw std::runtime_error("Unexpected operator");
        }
        advance();
        
        // Handle "by" after "divided by"
        if (op == BinaryOpNode::Op::Divide && currentToken().type == TokenType::IDENTIFIER && 
            currentToken().value == "by") {
            advance();
        }
        
        auto right = parsePrimary();
        left = std::make_unique<BinaryOpNode>(op, std::move(left), std::move(right));
    }
    
    return left;
}

std::unique_ptr<ASTNode> Parser::parsePrimary() {
    switch (currentToken().type) {
        case TokenType::NUMBER: {
            double value = std::stod(currentToken().value);
            advance();
            return std::make_unique<NumberNode>(value);
        }
        
        case TokenType::STRING: {
            std::string value = currentToken().value;
            advance();
            return std::make_unique<StringNode>(value);
        }
        
        case TokenType::IDENTIFIER: {
            std::string name = currentToken().value;
            advance();
            
            // Check if it's a function call
            if (currentToken().type == TokenType::LEFT_PAREN) {
                advance(); // Skip '('
                
                std::vector<std::unique_ptr<ASTNode>> args;
                
                // Parse arguments
                if (currentToken().type != TokenType::RIGHT_PAREN) {
                    args.push_back(parseExpression());
                    
                    while (currentToken().type == TokenType::COMMA) {
                        advance(); // Skip comma
                        args.push_back(parseExpression());
                    }
                }
                
                // Expect closing parenthesis
                if (currentToken().type != TokenType::RIGHT_PAREN) {
                    throw std::runtime_error("Expected ')' after function arguments");
                }
                advance(); // Skip ')'
                
                return std::make_unique<CallNode>(name, std::move(args));
            } else {
                // Otherwise it's a variable reference
                return std::make_unique<IdentifierNode>(name);
            }
        }
        
        case TokenType::LEFT_PAREN: {
            advance(); // Skip '('
            auto expr = parseExpression();
            
            if (currentToken().type != TokenType::RIGHT_PAREN) {
                throw std::runtime_error("Expected ')'");
            }
            advance(); // Skip ')'
            
            return expr;
        }
        
        default:
            throw std::runtime_error("Unexpected token: " + currentToken().value);
    }
}

std::unique_ptr<ASTNode> Parser::parseFunction() {
    advance(); // Skip 'function' keyword
    
    std::string name;
    if (currentToken().type == TokenType::IDENTIFIER) {
        name = currentToken().value;
        advance();
    } else {
        throw std::runtime_error("Expected function name");
    }
    
    // Parse parameter list
    if (currentToken().type != TokenType::LEFT_PAREN) {
        throw std::runtime_error("Expected '(' after function name");
    }
    advance(); // Skip '('
    
    std::vector<std::string> parameters;
    
    // Parse parameters
    if (currentToken().type != TokenType::RIGHT_PAREN) {
        if (currentToken().type != TokenType::IDENTIFIER) {
            throw std::runtime_error("Expected parameter name");
        }
        
        parameters.push_back(currentToken().value);
        advance();
        
        while (currentToken().type == TokenType::COMMA) {
            advance(); // Skip comma
            
            if (currentToken().type != TokenType::IDENTIFIER) {
                throw std::runtime_error("Expected parameter name after comma");
            }
            
            parameters.push_back(currentToken().value);
            advance();
        }
    }
    
    // Expect closing parenthesis
    if (currentToken().type != TokenType::RIGHT_PAREN) {
        throw std::runtime_error("Expected ')' after parameters");
    }
    advance(); // Skip ')'
    
    // Parse function body
    auto body = std::unique_ptr<BlockNode>(dynamic_cast<BlockNode*>(parseBlock().release()));
    if (!body) {
        throw std::runtime_error("Expected function body");
    }
    
    return std::make_unique<FunctionNode>(name, std::move(parameters), std::move(body));
}

std::unique_ptr<ASTNode> Parser::parseIfStatement() {
    advance(); // Skip 'if' keyword
    
    auto condition = parseExpression();
    
    auto thenBlock = std::unique_ptr<BlockNode>(dynamic_cast<BlockNode*>(parseBlock().release()));
    if (!thenBlock) {
        throw std::runtime_error("Expected block after if condition");
    }
    
    std::unique_ptr<BlockNode> elseBlock;
    if (currentToken().type == TokenType::ELSE) {
        advance(); // Skip 'else' keyword
        
        elseBlock = std::unique_ptr<BlockNode>(dynamic_cast<BlockNode*>(parseBlock().release()));
        if (!elseBlock) {
            throw std::runtime_error("Expected block after else");
        }
    }
    
    return std::make_unique<IfNode>(std::move(condition), std::move(thenBlock), std::move(elseBlock));
}

std::unique_ptr<ASTNode> Parser::parseWhileLoop() {
    advance(); // Skip 'while' keyword
    
    auto condition = parseExpression();
    
    auto body = std::unique_ptr<BlockNode>(dynamic_cast<BlockNode*>(parseBlock().release()));
    if (!body) {
        throw std::runtime_error("Expected block after while condition");
    }
    
    return std::make_unique<WhileNode>(std::move(condition), std::move(body));
}

std::unique_ptr<ASTNode> Parser::parsePrintStatement() {
    advance(); // Skip 'print' keyword
    
    auto expr = parseExpression();
    
    return std::make_unique<PrintNode>(std::move(expr));
}

File: src/parser.rs
use crate::ast::*;
use crate::token::{Token, TokenType};
use anyhow::{anyhow, Result};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, current: 0 }
    }

    fn current_token(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn peek_token(&self, offset: usize) -> Option<&Token> {
        let pos = self.current + offset;
        if pos < self.tokens.len() {
            Some(&self.tokens[pos])
        } else {
            None
        }
    }

    fn advance(&mut self) {
        if self.current < self.tokens.len() - 1 {
            self.current += 1;
        }
    }

    fn expect(&mut self, expected: &TokenType) -> Result<()> {
        if std::mem::discriminant(&self.current_token().token_type)
            == std::mem::discriminant(expected)
        {
            self.advance();
            Ok(())
        } else {
            Err(anyhow!(
                "Expected {:?}, found {:?} at line {}",
                expected,
                self.current_token().token_type,
                self.current_token().line
            ))
        }
    }

    pub fn parse(&mut self) -> Result<Program> {
        let mut functions = Vec::new();

        while !matches!(self.current_token().token_type, TokenType::Eof) {
            functions.push(self.parse_function()?);
        }

        Ok(Program { functions })
    }

    fn parse_function(&mut self) -> Result<Function> {
        self.expect(&TokenType::Function)?;

        let name = if let TokenType::Identifier(n) = &self.current_token().token_type {
            let name = n.clone();
            self.advance();
            name
        } else {
            return Err(anyhow!("Expected function name"));
        };

        self.expect(&TokenType::LeftParen)?;

        let mut params = Vec::new();
        while !matches!(self.current_token().token_type, TokenType::RightParen) {
            if let TokenType::Identifier(param) = &self.current_token().token_type {
                params.push(param.clone());
                self.advance();

                if matches!(self.current_token().token_type, TokenType::Comma) {
                    self.advance();
                }
            } else {
                return Err(anyhow!("Expected parameter name"));
            }
        }

        self.expect(&TokenType::RightParen)?;
        self.expect(&TokenType::LeftBrace)?;

        let body = self.parse_block()?;

        self.expect(&TokenType::RightBrace)?;

        Ok(Function { name, params, body })
    }

    fn parse_block(&mut self) -> Result<Vec<Stmt>> {
        let mut statements = Vec::new();

        while !matches!(
            self.current_token().token_type,
            TokenType::RightBrace | TokenType::Eof
        ) {
            statements.push(self.parse_statement()?);
        }

        Ok(statements)
    }

    fn parse_statement(&mut self) -> Result<Stmt> {
        match &self.current_token().token_type {
            TokenType::Variable => self.parse_var_decl(false),
            TokenType::Constant => self.parse_var_decl(true),
            TokenType::Print => self.parse_print(),
            TokenType::If => self.parse_if(),
            TokenType::While => self.parse_while(),
            TokenType::For => self.parse_for(),
            TokenType::Return => self.parse_return(),
            TokenType::Identifier(_) => self.parse_assignment_or_expr(),
            _ => Err(anyhow!(
                "Unexpected token: {:?} at line {}",
                self.current_token().token_type,
                self.current_token().line
            )),
        }
    }

    fn parse_var_decl(&mut self, is_const: bool) -> Result<Stmt> {
        self.advance(); // Skip 'var' or 'const'

        let name = if let TokenType::Identifier(n) = &self.current_token().token_type {
            let name = n.clone();
            self.advance();
            name
        } else {
            return Err(anyhow!("Expected variable name"));
        };

        self.expect(&TokenType::Assign)?;

        let value = self.parse_expression()?;

        Ok(Stmt::VarDecl {
            name,
            value,
            is_const,
        })
    }

    fn parse_assignment_or_expr(&mut self) -> Result<Stmt> {
        let name = if let TokenType::Identifier(n) = &self.current_token().token_type {
            let name = n.clone();
            self.advance();
            name
        } else {
            return Err(anyhow!("Expected identifier"));
        };

        if matches!(self.current_token().token_type, TokenType::Assign) {
            self.advance();
            let value = self.parse_expression()?;
            Ok(Stmt::Assignment { name, value })
        } else {
            // It's a function call
            self.current -= 1; // Go back
            let expr = self.parse_expression()?;
            Ok(Stmt::Expression(expr))
        }
    }

    fn parse_print(&mut self) -> Result<Stmt> {
        self.advance(); // Skip 'print'
        let expr = self.parse_expression()?;
        Ok(Stmt::Print(expr))
    }

    fn parse_if(&mut self) -> Result<Stmt> {
        self.advance(); // Skip 'if'

        let condition = self.parse_expression()?;
        self.expect(&TokenType::LeftBrace)?;
        let then_block = self.parse_block()?;
        self.expect(&TokenType::RightBrace)?;

        let else_block = if matches!(self.current_token().token_type, TokenType::Else) {
            self.advance();
            self.expect(&TokenType::LeftBrace)?;
            let block = self.parse_block()?;
            self.expect(&TokenType::RightBrace)?;
            Some(block)
        } else {
            None
        };

        Ok(Stmt::If {
            condition,
            then_block,
            else_block,
        })
    }

    fn parse_while(&mut self) -> Result<Stmt> {
        self.advance(); // Skip 'while'

        let condition = self.parse_expression()?;
        self.expect(&TokenType::LeftBrace)?;
        let body = self.parse_block()?;
        self.expect(&TokenType::RightBrace)?;

        Ok(Stmt::While { condition, body })
    }

    fn parse_for(&mut self) -> Result<Stmt> {
        self.advance(); // Skip 'for'

        self.expect(&TokenType::LeftParen)?;

        let init = Box::new(self.parse_statement()?);

        let condition = self.parse_expression()?;

        let increment = Box::new(self.parse_statement()?);

        self.expect(&TokenType::RightParen)?;
        self.expect(&TokenType::LeftBrace)?;
        let body = self.parse_block()?;
        self.expect(&TokenType::RightBrace)?;

        Ok(Stmt::For {
            init,
            condition,
            increment,
            body,
        })
    }

    fn parse_return(&mut self) -> Result<Stmt> {
        self.advance(); // Skip 'return'

        if matches!(
            self.current_token().token_type,
            TokenType::RightBrace | TokenType::Eof
        ) {
            Ok(Stmt::Return(None))
        } else {
            let expr = self.parse_expression()?;
            Ok(Stmt::Return(Some(expr)))
        }
    }

    fn parse_expression(&mut self) -> Result<Expr> {
        self.parse_or()
    }

    fn parse_or(&mut self) -> Result<Expr> {
        let mut left = self.parse_and()?;

        while matches!(self.current_token().token_type, TokenType::Or) {
            self.advance();
            let right = self.parse_and()?;
            left = Expr::Binary {
                left: Box::new(left),
                op: BinaryOp::Or,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Expr> {
        let mut left = self.parse_equality()?;

        while matches!(self.current_token().token_type, TokenType::And) {
            self.advance();
            let right = self.parse_equality()?;
            left = Expr::Binary {
                left: Box::new(left),
                op: BinaryOp::And,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_equality(&mut self) -> Result<Expr> {
        let mut left = self.parse_comparison()?;

        loop {
            let op = match &self.current_token().token_type {
                TokenType::Equals => BinaryOp::Equals,
                _ => break,
            };

            self.advance();
            let right = self.parse_comparison()?;
            left = Expr::Binary {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_comparison(&mut self) -> Result<Expr> {
        let mut left = self.parse_term()?;

        loop {
            let op = match &self.current_token().token_type {
                TokenType::Greater => {
                    self.advance();
                    // Check for "greater than"
                    if let Some(TokenType::Identifier(ref s)) =
                        self.peek_token(0).map(|t| &t.token_type)
                    {
                        if s == "than" {
                            self.advance();
                        }
                    }
                    BinaryOp::Greater
                }
                TokenType::Less => {
                    self.advance();
                    // Check for "less than"
                    if let Some(TokenType::Identifier(ref s)) =
                        self.peek_token(0).map(|t| &t.token_type)
                    {
                        if s == "than" {
                            self.advance();
                        }
                    }
                    BinaryOp::Less
                }
                _ => break,
            };

            let right = self.parse_term()?;
            left = Expr::Binary {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_term(&mut self) -> Result<Expr> {
        let mut left = self.parse_factor()?;

        loop {
            let op = match &self.current_token().token_type {
                TokenType::Plus => BinaryOp::Add,
                TokenType::Minus => BinaryOp::Subtract,
                _ => break,
            };

            self.advance();
            let right = self.parse_factor()?;
            left = Expr::Binary {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_factor(&mut self) -> Result<Expr> {
        let mut left = self.parse_unary()?;

        loop {
            let op = match &self.current_token().token_type {
                TokenType::Multiply => BinaryOp::Multiply,
                TokenType::Divide => BinaryOp::Divide,
                _ => break,
            };

            self.advance();
            let right = self.parse_unary()?;
            left = Expr::Binary {
                left: Box::new(left),
                op,
                right: Box::new(right),
            };
        }

        Ok(left)
    }

    fn parse_unary(&mut self) -> Result<Expr> {
        match &self.current_token().token_type {
            TokenType::Not => {
                self.advance();
                let operand = self.parse_unary()?;
                Ok(Expr::Unary {
                    op: UnaryOp::Not,
                    operand: Box::new(operand),
                })
            }
            TokenType::Minus => {
                self.advance();
                let operand = self.parse_unary()?;
                Ok(Expr::Unary {
                    op: UnaryOp::Negate,
                    operand: Box::new(operand),
                })
            }
            _ => self.parse_primary(),
        }
    }

    fn parse_primary(&mut self) -> Result<Expr> {
        match &self.current_token().token_type.clone() {
            TokenType::Number(n) => {
                let num = *n;
                self.advance();
                Ok(Expr::Number(num))
            }
            TokenType::StringLit(s) => {
                let string = s.clone();
                self.advance();
                Ok(Expr::String(string))
            }
            TokenType::Boolean(b) => {
                let val = *b;
                self.advance();
                Ok(Expr::Boolean(val))
            }
            TokenType::Identifier(name) => {
                let name = name.clone();
                self.advance();

                // Check for function call
                if matches!(self.current_token().token_type, TokenType::LeftParen) {
                    self.advance();
                    let mut args = Vec::new();

                    while !matches!(self.current_token().token_type, TokenType::RightParen) {
                        args.push(self.parse_expression()?);

                        if matches!(self.current_token().token_type, TokenType::Comma) {
                            self.advance();
                        }
                    }

                    self.expect(&TokenType::RightParen)?;
                    Ok(Expr::Call { name, args })
                } else {
                    Ok(Expr::Identifier(name))
                }
            }
            TokenType::LeftParen => {
                self.advance();
                let expr = self.parse_expression()?;
                self.expect(&TokenType::RightParen)?;
                Ok(expr)
            }
            _ => Err(anyhow!(
                "Unexpected token in expression: {:?} at line {}",
                self.current_token().token_type,
                self.current_token().line
            )),
        }
    }
}
